rep-0149.rst

```diff
@@ -0,0 +1,1047 @@
+
+  <export>
+    <message_generator>cpp</message_generator>
+  </export>
+
+<metapackage/>
+''''''''''''''
+
+This empty tag declares a special kind of catkin package used for
+grouping other packages.  Metapackages only provide execution-time
+dependencies.  They cannot be used for catkin builds and compile
+nothing themselves.  Metapackages may not install any code or other
+files, although ``package.xml`` does get installed automatically.
+They can depend on other metapackages, but regular catkin packages
+cannot.
+
+A good use for metapackages is to group the major components of your
+robot and then provide a comprehensive grouping for your whole system.
+Package installation tools like ``apt-get`` or ``yum`` can
+automatically install all the packages on which a metapackage directly
+or indirectly depends.  Metapackages can also be used to resolve
+dependencies declared by legacy rosbuild stacks not yet converted to
+catkin.
+
+Every metapackage must have a ``CMakeLists.txt`` containing these
+commands::
+
+  cmake_minimum_required(VERSION 2.8.3)
+  project(PACKAGE_NAME)
+  find_package(catkin REQUIRED)
+  catkin_metapackage()
+
+Because the metapackage ``CMakeLists.txt`` contains a catkin macro,
+its ``package.xml`` must declare a buildtool dependency on catkin::
+
+  <buildtool_depend>catkin</buildtool_depend>
+
+Additional buildtool, build or test dependencies are not permitted.
+
+Because metapackages only supply execution-time dependencies, they use
+``<exec_depend>`` to list the packages in their group::
+
+  <exec_depend>your_custom_msgs</exec_depend>
+  <exec_depend>your_server_node</exec_depend>
+  <exec_depend>your_utils</exec_depend>
+  <exec_depend>another_metapackage</exec_depend>
+
+.. raw:: html
+
+  <font color="red">
+
+<member_of_group> (multiple)
+''''''''''''''''''''''''''''
+
+The content identifies the name of the dependency group::
+
+  <export>
+    <member_of_group>message_packages</member_of_group>
+  </export>
+
+.. raw:: html
+
+  </font>
+
+
+Compatibility
+=============
+
+.. raw:: html
+
+  <font color="red">
+
+Support for format three
+------------------------
+
+Existing tools supporting up to format two will need to be updated to
+honor the new information provided by format three.
+If they are not checking that the format is two or lower they will
+simply ignore the format three specific information when processing a
+package with a format three manifest.
+
+catkin_prepare_release
+''''''''''''''''''''''
+
+As long as a manifest doesn't specify an `abi_version` the tools doesn't
+need to change.
+For packages which do specify an `abi_version` the tool should probably
+by default remove the attribute and only after confirmation from the user
+offer to keep it.
+
+ros_buildfarm
+'''''''''''''
+
+The dependencies between packages is directly mapped to upstream /
+downstream jobs in Jenkins.
+In order to consider the `abi_version` attribute these explicit job
```

Comment:
We could probably keep the dependencies but have a shortcut/check for ABI compatibility at the top of the build and it will just pass without a rebuild if an ABI compatible version already exists.

Reply:
This would require downstream jobs to be triggered just in order to determine that they should not be rebuild. I would rather avoid that overhead and not trigger any downstream jobs in the first place.

Reply:
We either have to keep the explicit job ordering and short circuit for ABI compatibility switch to programatically scripted jobs and deal with potential race conditions when a package build is triggered by one upstream build but other upstreams are still building.

I would guess that the overhead of queuing and short circuiting is less than trying to iron out all of the behavior of a system where we write our own triggers.

It's also possible that there may be options or plugins that allow us to keep the explicit dependency relationship between jobs in Jenkins but only trigger jobs in certain conditions.